Matrix documentation (#2694)
jenkins-infra/jenkins.io: content/doc/book/pipeline/syntax.adoc

}// Script //
----

[[declarative-matrix]]
=== Matrix

Stages in Declarative Pipeline may have a `matrix` section defining a multi-dimenisional matrix of name-value combinations to be run in parallel. 
We'll refer these combinations as "cells" in a matrix.

> Each cell in a matrix can include a one or stages to be run in sequential order using the configuration for that cell.

Reviewer: ```suggestion
Each cell in a matrix can include one or more stages to be run sequentially using the configuration for that cell.
```

Note that a stage must have one and only one of `steps`, `stages`, `parallel`, or `matrix`. 
It is not possible to nest a `parallel` or `matrix` block within a `stage` directive if that `stage`
directive is nested within a `parallel`  or `matrix` block itself. However, a `stage`
directive within a `parallel` or `matrix` block can use all other functionality of a `stage`,
including `agent`, `tools`, `when`, etc.

In addition, you can force your `matrix` cells to all be aborted when any one
of them fails, by adding `failFast true` to the `stage` containing the
`matrix`. Another option for adding `failfast` is adding an option to the
pipeline definition: `parallelsAlwaysFailFast()`

> The `matrix` section must include an `axes` section defining matrix of 

Reviewer: I don't understand this phrasing
```suggestion
The `matrix` section must include an `axes` section defining matrix
```

and a `stages` section defining stages to run sequentially in that cell.
A `matrix` may have an `excludes` section to filter out invalid cells from the matrix.

> Many of directives available on  `stage`, including `agent`, `tools`, `when`, etc., 

Reviewer: ```suggestion
Many of the directives available on  `stage`, including `agent`, `tools`, `when`, etc., 
```

All the values from each axis are combined with the others to produce the cells. 

[[matrix-axes-example]]
===== Example

.One-axis with 3 cells
[source,groovy]
----
matrix {
    axes {
        axis {
            name 'PLATFORM'
            values 'linux', 'mac', 'windows' 
        }
    }
    // ...
}
----

.Two-axis with 12 cells (three by four)
[source, groovy]
----
matrix {
    axes {
        axis {
            name 'PLATFORM'
            values 'linux', 'mac', 'windows' 
        }
        axis {
            name 'BROWSER'
            values 'chrome', 'edge', 'firefox', 'safari' 
        }
    }
    // ...
}
----

.Three-axis matrix with 24 cells (three by four by two)
[source,groovy]
----
matrix {
    axes {
        axis {
            name 'PLATFORM'
            values 'linux', 'mac', 'windows' 
        }
        axis {
            name 'BROWSER'
            values 'chrome', 'edge', 'firefox', 'safari' 
        }
        axis {
            name 'ARCHITECTURE'
            values '32-bit', '64-bit'
        }
    }
    // ...
}
----

[[matrix-stages]]
==== stages   

The `stages` section specifies one or more `stage`s to be executed sequentially in each cell. 

> This is section is identical to any other

Reviewer: ```suggestion
This section is identical to any other
```
